### PR TITLE
Phantom bridge

### PR DESCRIPTION
**Pull request by keraldi**
Rework.
Diagonally doesn't work properly yet, some variables are switched up which is why the template doesn't work properly.
Rest works and should work better than before.
Just cannot go speeding diagonally.

### CODE REVIEW

**Comment by keraldi** [comment]
~ finished please review ~

File: src/main/java/WayofTime/bloodmagic/item/sigil/ItemSigilPhantomBridge.java [hunk 1]
```diff
@@ -1,19 +1,74 @@
...
+                    if (invStack.getItem() instanceof ItemSigilPhantomBridge)
+                        return NBTHelper.checkNBT(invStack).getTagCompound().getBoolean(Constants.NBT.ACTIVATED);
+                }
+            }
+
+        }
+        return false;
+    }
+
+    public static final String[] circle7x7 = new String[]{
```
Comment (TehNut) [inline src/main/java/WayofTime/bloodmagic/item/sigil/ItemSigilPhantomBridge.java]: Constants should be `CONSTANT_CASE` and should appear at the top of the class.

File: src/main/java/WayofTime/bloodmagic/item/sigil/ItemSigilPhantomBridge.java [hunk 2]
```diff
@@ -1,19 +1,74 @@
...
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 
 public class ItemSigilPhantomBridge extends ItemSigilToggleableBase {
-    private static final double EXPANSION_FACTOR = 2;
-    private static final int RANGE = 3;
+    public static final Predicate<Entity> IS_PHANTOM_ACTIVE = (Entity entity) -> entity instanceof EntityPlayer && isPhantomActive((EntityPlayer) entity);
+
+    public static boolean isPhantomActive(EntityPlayer entity) {
```
Comment (TehNut) [inline src/main/java/WayofTime/bloodmagic/item/sigil/ItemSigilPhantomBridge.java]: Static methods should appear below all member methods.

File: src/main/java/WayofTime/bloodmagic/item/sigil/ItemSigilPhantomBridge.java [hunk 3]
```diff
@@ -1,19 +1,74 @@
...
+            "  XXX  ",
+            " XXXXX ",
+            "XXXXXXX",
+            "XXXXXXX",
+            "XXXXXXX",
+            " XXXXX ",
+            "  XXX  "
+    };
+
+    public static final String[] circle9x9 = new String[]{
```
Comment (TehNut) [inline src/main/java/WayofTime/bloodmagic/item/sigil/ItemSigilPhantomBridge.java]: Constants should be `CONSTANT_CASE` and should appear at the top of the class.

File: src/main/java/WayofTime/bloodmagic/item/sigil/ItemSigilPhantomBridge.java [hunk 4]
```diff
@@ -1,19 +1,74 @@
...
+            "XXXXXXXXX",
+            "XXXXXXXXX",
+            "XXXXXXXXX",
+            "XXXXXXXXX",
+            " XXXXXXX ",
+            "  XXXXX  "
+    };
+
+    //imagine you're looking into positive Z
+    public static final String[] diagXZ = new String[] {
```
Comment (TehNut) [inline src/main/java/WayofTime/bloodmagic/item/sigil/ItemSigilPhantomBridge.java]: Constants should be `CONSTANT_CASE` and should appear at the top of the class.

File: src/main/java/WayofTime/bloodmagic/item/sigil/ItemSigilPhantomBridge.java [hunk 5]
```diff
@@ -58,32 +114,146 @@ public void onSigilUpdate(ItemStack stack, World world, EntityPlayer player, int
...
-        double offsetPosX = posX + EXPANSION_FACTOR * playerVelX;
-        double offsetPosZ = posZ + EXPANSION_FACTOR * playerVelZ;
-        double avgX = (posX + offsetPosX) / 2;
-        double avgZ = (posZ + offsetPosZ) / 2;
+        //Standing still, sneaking or walking with framerate drops
+        if (0 <= totalVel && totalVel < 0.1) {
+            circleTemplate7x7(posX, posY, posZ, verticalOffset, world);
+        } else //anything between the first case and being slightly faster than walking
+            //walking fairly quickly on X-axis
+            if (-0.3 < playerVelZ && playerVelZ < 0.3) {
```
Comment (TehNut) [inline src/main/java/WayofTime/bloodmagic/item/sigil/ItemSigilPhantomBridge.java]: This reads very strangely. It should be `playerVelZ > -0.3 && playerVelZ < 0.3`. I understand you're trying to show a range, but it just makes it more confusing.

Same for all the other ranges done this way.

File: src/main/java/WayofTime/bloodmagic/item/sigil/ItemSigilPhantomBridge.java [hunk 6]
```diff
@@ -58,32 +114,146 @@ public void onSigilUpdate(ItemStack stack, World world, EntityPlayer player, int
...
+                    rectangleTemplatePosZ(posX, posY, posZ, verticalOffset, world);
+                }
+                if (playerVelZ < -0.3) {
+                    if (playerVelZ < -1) {
+                        rectangleTemplateNegZLong(posX, posY, posZ, verticalOffset, world);
+                    }
+                    rectangleTemplateNegZ(posX, posY, posZ, verticalOffset, world);
+                }
+            }
+            else //diagonal movement
```
Comment (TehNut) [inline src/main/java/WayofTime/bloodmagic/item/sigil/ItemSigilPhantomBridge.java]: This should have braces. It should also be on the same line as the brace above.

File: src/main/java/WayofTime/bloodmagic/item/sigil/ItemSigilPhantomBridge.java [hunk 7]
```diff
@@ -58,32 +114,146 @@ public void onSigilUpdate(ItemStack stack, World world, EntityPlayer player, int
...
         int posZ = playerPos.getZ();
 
-        double offsetPosX = posX + EXPANSION_FACTOR * playerVelX;
-        double offsetPosZ = posZ + EXPANSION_FACTOR * playerVelZ;
-        double avgX = (posX + offsetPosX) / 2;
-        double avgZ = (posZ + offsetPosZ) / 2;
+        //Standing still, sneaking or walking with framerate drops
+        if (0 <= totalVel && totalVel < 0.1) {
+            circleTemplate7x7(posX, posY, posZ, verticalOffset, world);
+        } else //anything between the first case and being slightly faster than walking
```
Comment (TehNut) [inline src/main/java/WayofTime/bloodmagic/item/sigil/ItemSigilPhantomBridge.java]: This needs a brace.

File: src/main/java/WayofTime/bloodmagic/item/sigil/ItemSigilPhantomBridge.java [hunk 8]
```diff
@@ -58,32 +114,146 @@ public void onSigilUpdate(ItemStack stack, World world, EntityPlayer player, int
...
+        }
     }
+    private static void templateReaderDiag(int posX, int posY, int posZ, int verticalOffset, World world, int offsetZ, int offsetX, String[] template, boolean inverted, boolean XnZ) {
+        int i = 0;
+        for (String a: template) {
+            if(inverted) i--;
+            else i++;
+            int j = 0;
+            for (char b: a.toCharArray()) {
+                if(inverted && !XnZ || XnZ && !inverted) j--;
```
Comment (TehNut) [inline src/main/java/WayofTime/bloodmagic/item/sigil/ItemSigilPhantomBridge.java]: Newlines for this ifelse

**Review by TehNut — CHANGES_REQUESTED** [review]
Initial pass. I'm too tired to actually go over the logic right now.

I'm not a huge fan of all those 1 line helper methods.

**Comment by keraldi** [comment]
Done.

**Comment by keraldi** [comment]
Alternative explaination: I wanted more code.

**Comment by TehNut** [comment]
My brain is too fried to try and understand this. If this doesn't work, you're toast.

**Review by TehNut — APPROVED** [review]

**Pull request merged**